### PR TITLE
Hashable and Immutable ParamSpec

### PR DESCRIPTION
**Pull request by astafan8**
Introduction of `__eq__` for `ParamSpec` introduced implicit `__hash__ = None` definition into the class (according to Python 3 rules) which cause some rare code that was making `set`s of `ParamSpec`s break.

This PR adds `__hash__` method to `ParamSpec` that allows to create use `ParamSpec`s as other hashable objects, for example, making sets of them.

Moreover, `ParamSpec` is made immutable. `add_depends_on` and `add_inferred_from` methods have been removed.

Also, `depends_on_` and `inferred_from_` properties (with `_` at the end) have been added which return a list of strings as opposed to a string with comma-separated names.

- [x] change pytopo/sweep to not use `add_depends_on` and `add_inferred_from` methods

### CODE REVIEW

**Comment by WilliamHPNielsen** [comment]
> What is dangerous, however, is that ParamSpec is still a mutable object via the add_depends_on and add_inferred_from methods. In some earlier PR it was decided to get rid of those soon.

I think it could be made part of this PR to remove those two methods.

**Comment by sohailc** [comment]
> > What is dangerous, however, is that ParamSpec is still a mutable object via the add_depends_on and add_inferred_from methods. In some earlier PR it was decided to get rid of those soon.
> 
> I think it could be made part of this PR to remove those two methods.

The `add_depends_on` and `add_inferred_from` are used in pysweep. See: https://github.com/kouwenhovenlab/pytopo/blob/master/pytopo/sweep/param_table.py 

line 97. If you want a working pysweep @astafan8, do not remove those!

File: qcodes/dataset/param_spec.py [hunk 1]
```diff
@@ -105,6 +105,26 @@ def __eq__(self, other):
...
+
+        # First, get the hash of the tuple with all the relevant attributes
+        all_attr_tuple_hash = hash(
+            tuple(getattr(self, attr) for attr in attrs_with_strings)
+            + tuple(tuple(getattr(self, attr)) for attr in attrs_with_lists)
+        )
+        hash_value = all_attr_tuple_hash
+
+        # Then, XOR it with the individual hashes of all relevant attributes
+        for attr in attrs_with_strings:
```
[the remark this answers is not in the document]
Comment (astafan8) [reply qcodes/dataset/param_spec.py]: it pops up within the first 10 google search items when you look for "creating hash in python" or similar :) and the comments there seem convincing.

**Comment by astafan8** [comment]
@QCoDeS/core ready for review

**Comment by jenshnielsen** [comment]
Looks good but there seems to be some intermittent issue with the hypothesis generation

```
    @given(paramspecs=hst.lists(valid_paramspec_kwargs, min_size=6, max_size=6),
>          add_to_1_inf=hst.booleans(),
           add_to_1_dep=hst.booleans(),
           add_to_2_inf=hst.booleans(),
           add_to_2_dep=hst.booleans(),
           )
    def test_hash_with_deferred_and_inferred_as_paramspecs(
            paramspecs, add_to_1_inf, add_to_1_dep, add_to_2_inf, add_to_2_dep):
        """
        Test that hashing works if 'inferred_from' and/or 'depend_on' contain
        actual ParamSpec instances and not just strings.
E       hypothesis.errors.FailedHealthCheck: Data generation is extremely slow: Only produced 9 valid examples in 1.07 seconds (0 invalid ones and 0 exceeded maximum size). Try decreasing size of the data you're generating (with e.g.max_size or max_leaves parameters).
E       See https://hypothesis.readthedocs.io/en/latest/healthchecks.html for more information about this. If you want to disable just this health check, add HealthCheck.too_slow to the suppress_health_check settings for this test.
```

**Comment by astafan8** [comment]
cool. may we give it an extra week? (the sweep has been updated some days ago, users have been informed, but not sure that everyone was updated) unless, this blocks the "interdependencies refactoring PR" very prominently.

**Comment by jenshnielsen** [comment]
Sounds like a good idea, merging this on a Friday is probably not the best idea

**Comment by jenshnielsen** [comment]
@astafan8 Should we merge this?

**Pull request merged**